src/lib/transition_handler/breadcrumb_builder.ml

```diff
@@ -90,24 +89,50 @@ module Make (Inputs : Inputs.S) :
                            breadcrumb_if_present ()
                          in
                          new_breadcrumb)
-                  | Error (`Fatal_error exn) ->
-                      Deferred.return (Or_error.of_exn exn)
-                  | Error (`Invalid_staged_ledger_hash error) ->
-                      let%map () =
-                        Trust_system.record_envelope_sender trust_system logger
-                          sender
-                          ( Trust_system.Actions.Gossiped_invalid_transition
-                          , Some ("invalid staged ledger hash", []) )
+                  | Error err -> (
+                      (* propagate bans through subtrees *)
+                      let subtrees =
```

Comment:
This is the entire set of subtrees we are attempting to add, right? If so, this is incorrect. We only want to punish people who sent us blocks which are children of a bad block that we found. If there are parent blocks that were valid before this, we don't want to punish peers who sent only those since they could have a valid view of the network and not be acting adversarial. When we find a bad block in the any of the subtrees we are adding, we should punish the peer that sent that and any peers that send child blocks and then continue working on the other subtrees (adding as much as we can). We should still return an error the the caller, but it's best to waste as little good information as possible, ignoring just the adversarial information.